apps_download/models/product_product.py

```diff
@@ -0,0 +1,113 @@
+
+import os
+import tempfile
+import shutil
+import logging
+import base64
+import subprocess
+import time
+from odoo.exceptions import ValidationError
+from odoo import models, fields, api, _
+_logger = logging.getLogger(__name__)
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    dependent_product_ids = fields.Many2many(
+        'product.product', 'product_product_dependent_rel',
+        'src_id', 'dest_id', string='Dependent Products'
+    )
+    module_path = fields.Char('Module Path')
+
+    @api.model
+    def child_dependency_check(self, product_dependent_ids, children):
+        res = self.env['product.product']
+        for child in children:
+            if not child.dependent_product_ids:
+                continue
+            if child in product_dependent_ids:
+                raise ValidationError(
+                    _('Error: You cannot create recursive dependency.')
+                )
+            product_dependent_ids += child
+            self.child_dependency_check(product_dependent_ids,
+                                        child.dependent_product_ids)
+        return res
+    @api.constrains('dependent_product_ids')
+    def check_dependent_recursion(self):
+        for product in self:
+            self.child_dependency_check(product, product.dependent_product_ids)
+
+    @api.model
+    def child_dependency(self, children):
+        res = self.env['product.product']
+        for child in children:
+            if not child.dependent_product_ids:
+                continue
+            res += child.dependent_product_ids
+            res += self.child_dependency(child.dependent_product_ids)
+        return res
+
+    @api.multi
+    def create_dependency_list(self):
+        ret_val = {}
+        for product in self:
+            ret_val[product.id] = product.dependent_product_ids
+            if product.dependent_product_ids:
+                ret_val[product.id] += self.child_dependency(
+                    product.dependent_product_ids)
+        return ret_val
+
+    @api.multi
+    def generate_zip_file(self):
+        for product in self:
+            if not product.module_path:
+                continue
+            tmp_dir = tempfile.mkdtemp()
+            tmp_dir_2 = tempfile.mkdtemp()
+            dependent_products = product.create_dependency_list()
+            dependent_products = dependent_products[product.id]
+            for dependent_product in dependent_products:
+                if not dependent_product.module_path:
+                    continue
+                subprocess.Popen(['cp', '-r', dependent_product.module_path,
+                                  tmp_dir], stdout=subprocess.PIPE)
+            subprocess.Popen(['cp', '-r', product.module_path, tmp_dir],
+                             stdout=subprocess.PIPE)
+            time_value = time.strftime(
+                '_%y%m%d_%H%M%S')
+
+            tmp_zip_file = os.path.join(tmp_dir_2, product.name) + time_value
```

Comment:
Would it make sense to add the version of the product in the name?
something like product_product-10.0.1.1.1-20171016.zip?